modules/tensor_mechanics/src/materials/FiniteStrainHyperElasticViscoPlastic.C

```diff
@@ -0,0 +1,394 @@
+  params.addRequiredParam<UserObjectName>("stress_user_object","User object that computes stress using elastic RightCauchyGreenTensor");
+  params.addRequiredParam<std::vector<UserObjectName> >("flow_rate_user_objects","List of User object names that computes flow rate and derivatives");
+
+  return params;
+}
+
+FiniteStrainHyperElasticViscoPlastic::FiniteStrainHyperElasticViscoPlastic(const std::string & name,
+                                                                           InputParameters parameters) :
+    FiniteStrainMaterial(name, parameters),
+  _resid_abs_tol(getParam<Real>("resid_abs_tol")),
+  _resid_rel_tol(getParam<Real>("resid_rel_tol")),
+  _maxiters(getParam<unsigned int>("maxiters")),
+  _max_substep_iter(getParam<unsigned int>("max_substep_iteration")),
+  _stress_uo(getUserObject<HyperElasticStress>("stress_user_object")),
+  _num_flow_rate_uos(getParam<std::vector<UserObjectName> >("flow_rate_user_objects").size()),
+  _fp(declareProperty<RankTwoTensor>("fp")),
+  _fp_old(declarePropertyOld<RankTwoTensor>("fp")),
+  _q(declareProperty<std::vector<Real> >("internal_var")),
+  _q_old(declarePropertyOld<std::vector<Real> >("internal_var")),
+  _deformation_gradient_old(declarePropertyOld<RankTwoTensor>("deformation gradient"))
+{
+  if (_num_flow_rate_uos == 0)
+    mooseError("Specify atleast one flow rate user object");
+
+  _flow_rate_uo.resize(_num_flow_rate_uos);
+  _q_end_index.resize(_num_flow_rate_uos, 0);
+
+  UserObjectInterface uoi(parameters);
+
+  _num_internal_var = 0;
+  for (unsigned int i = 0; i < _num_flow_rate_uos; ++i)
+  {
+    _flow_rate_uo[i] = &uoi.getUserObjectByName<FlowRateModel>(parameters.get< std::vector<UserObjectName> >("flow_rate_user_objects")[i]);
+    _num_internal_var += _flow_rate_uo[i]->numInternalVar();
+    _q_end_index[i] = _num_internal_var;
+  }
+
+  _q_tmp.resize(_num_internal_var, 0.0);
+  _q_tmp_old.resize(_num_internal_var, 0.0);
+
+  _resid.resize(_num_flow_rate_uos, 0.0);
+  _jac.resize(_num_flow_rate_uos * _num_flow_rate_uos, 0.0);
+
+  _flow_rate.resize(_num_flow_rate_uos, 0.0);
+  _flow_rate_func.resize(_num_flow_rate_uos, 0.0);
+  _flow_dirn.resize(_num_flow_rate_uos);
+  _dpk2_dflowrate.resize(_num_flow_rate_uos);
+
+  _dq_dflowrate.resize(_num_internal_var, 0.0);
+  _dflowrate_dpk2.resize(_num_flow_rate_uos);
+  _dflowrate_dq.resize(_num_internal_var, 0.0);
+
+  _iden.zero();
+  _iden.addIa(1.0);
+
+  _err_flag = true;
+}
+
+void
+FiniteStrainHyperElasticViscoPlastic::initQpStatefulProperties()
+{
+  _stress[_qp].zero();
+
+  _fp[_qp].zero();
+  _fp[_qp].addIa(1.0);
+
+  _q[_qp].resize(_num_internal_var, 0.0);
+  _q_old[_qp].resize(_num_internal_var, 0.0);
+}
+
+void
+FiniteStrainHyperElasticViscoPlastic::computeQpStress()
+{
+  if (_t_step == 0) return;
+
+  RankTwoTensor delta_dfgrd = _deformation_gradient[_qp] - _deformation_gradient_old[_qp];
+  unsigned int num_substep = 1;
+  unsigned int substep_iter = 1;
+
+  _err_flag = true;
```

Comment:
I'm not a big fan of state flags like these. Especially here the design where you set the error flag to true initially seems odd. In general state flags used to break out of nested loops indicate that you should factor loop bodies out into separat efunctions (which you can break out of with return! and the return value could be the error state).